### PR TITLE
add ru loc.

### PR DESCRIPTION
**Pull request by shikulja**
Attention: Check the functionality of the turtle-wow module before accepting.

### CODE REVIEW

**Comment by shikulja** [comment]
ideas:
1.  Add width variables to the localization configuration

image
image

2. The names of the categories are statistical. They can be included in the common core language pack

**Comment by shagu** [comment]
Hi, I had a look at this, and I think it would be better to have a dedicated table for all UI translations.

This would have the benefit that we can put a metatable on it, so we don't have to add values for strings that are required from other extensions like ShaguTweaks-mods. It would just return its index if no translation exists. (Same as in the pfUI localizations)

Same strings would also need to be translated just once and not in every file thats using it. 
Speaking of which, I don't like the idea to have translations in each and every module file. 
I think that the module file should be for logic, and not to be bloated with translations of all languages.

Beside of that, I really like the idea of the localization efforts. 
Please let me know if I can help with anything on it beside of givin my feedback.

For the automatic UI-sizing part, i think we can delay that and first focus on the translation.
So that we can handle the sizing issues in another PR after this here is done :)

In regards to the suggested metatable, please have a look here as example:
<https://github.com/shagu/pfUI/blob/master/pfUI.lua#L229-L233>

```lua
    pfUI.env.T = setmetatable(pfUI_translation[lang] or {}, { __index = function(tab,key)
      local value = tostring(key)
      rawset(tab,key,value)
      return value
    end})
```

It puts a function to the table, that whenever you try to access it, it will return its key as a fallback.
So when your table looks like this:
```lua
foo = {
  ["Hello"] = "Greetings",
}
```

It will return "Greetings" if you do `print(foo["Hello"]` and will return "Good Afternoon" if you do `print(foo["Good Afternoon"]`.
It returns the key it got accessed with, if it cannot find an entry in the table.

In short:
- Move translations in one file, one table.
- Use a metatable, to make sure we have translations for everything and don't need to hardcode thirdparty values to not break them.
- I think the `T[]` is good here, so we have a clear separation of client/code related translations and the visible ui text-strings.
- Ignore the UI-Sizing for now, and let's get the initial localization done first :)

Also, good job and thanks for your PR :)

**Comment by shikulja** [comment]
@shagu can I use L and T globally?
--[[
    local lang = GetLocale()
    ShaguTweaks.T = setmetatable(ShaguTweaks_translation[lang] or {}, { __index = function(tab,key)
      local value = tostring(key)
      rawset(tab,key,value)
      return value
    end})
--]]
local lang = GetLocale()
local locale = ShaguTweaks_locale[lang] or {}
local translation = ShaguTweaks_translation[lang] or {}

ShaguTweaks.L = setmetatable(locale, { __index = function(tab,key)
  local value = tostring(key)
  rawset(tab, key, value)
  return value
end})

ShaguTweaks.T = setmetatable(translation, { __index = ShaguTweaks.L })

T = ShaguTweaks.T or {}
ShaguTweaks.L = ShaguTweaks.L or {}

**Comment by shagu** [comment]
No, better do it local in each module like:
```lua
local L, T = ShaguTweaks.L, ShaguTweaks.T
```

**Comment by shikulja** [comment]
It will be a good idea to change the addon structure in the future, just like in pfui. That is, use init\xml

File: helpers.lua [hunk 1]
```diff
@@ -24,6 +24,26 @@ end
...
+local translation = ShaguTweaks_translation[lang] or {}
+
+ShaguTweaks.L = setmetatable(locale, { __index = function(tab,key)
+  local value = tostring(key)
+  rawset(tab, key, value)
+  return value
+end})
+
+ShaguTweaks.T = setmetatable(translation, { __index = ShaguTweaks.L })
+
```
Comment (shagu) [inline helpers.lua:46]: Putting that meta table on the `locales` could break things. I think to put that into the main.lua and only on the translation table would be better:

```diff
diff --git a/main.lua b/main.lua
index 1d5c4ff..bfa4d86 100644
--- a/main.lua
+++ b/main.lua
@@ -18,6 +18,15 @@ ShaguTweaks.mods = {}
 
 -- load translation tables
 ShaguTweaks.L = (ShaguTweaks_locale[GetLocale()] or ShaguTweaks_locale["enUS"])
+ShaguTweaks.T = (ShaguTweaks_translation[GetLocale()] or ShaguTweaks_translation["enUS"])
+
+-- use table index key as translation fallback
+ShaguTweaks.T = setmetatable(ShaguTweaks.T, { __index = function(tab,key)
+  local value = tostring(key)
+  rawset(tab, key, value)
+  return value
+end})
+
 ShaguTweaks:RegisterEvent("VARIABLES_LOADED")
 ShaguTweaks:SetScript("OnEvent", function()
   -- load current expansion
```

As a fictional example, checking if the nameplate belongs to a critter and should be hidden, one could do
if L["critters"][nameplate:GetName()] then nameplate:Hide() end

If that metatable would be set on the `L` it would always return true, as it finds **everything** as it always returns the index when there is nothing. Checks like this happen quite often on those tables, so I'd like to not have the metatable on it :)

Comment (shikulja) [reply helpers.lua:46]: Will you make changes? or should I make a commit? I don't know how Review mode works.

Comment (shagu) [reply helpers.lua:46]: > I don't know how Review mode works.

Me neither, I didn't meant to start a review mode, just wanted to comment on where i spotted it.
You do the commit please.

**Comment by shagu** [comment]
> It will be a good idea to change the addon structure in the future, just like in pfui. That is, use init\xml also modules sounds better instead mods for directory
> 
> in custom.lua can be add how to use localization (not me -P)

That's a good point. I will clean up the .toc files up once this PR is resolved.

**Comment by shagu** [comment]
Hi, thank you very much. Good job!
The PR is now merged via: https://github.com/shagu/ShaguTweaks/commit/fd56ed91927b14dbeb2009c8046443bac35afdc1
I had to rebase a bit to resolve the merge conflicts, so github didn't detect it.

Also I fixed the Turtle module (https://github.com/shagu/ShaguTweaks/commit/c97c7a4555ec1e02c07c43d250db719b73c4a0d8) where you by accident used a function call instead of the table index :)

**Comment by shikulja** [comment]
@shagu 
a line was missing
["Advanced Options"] = "Расширенные настройки",
in all translations, config.lua is not added to the parser?

if you decide to update, you can fix these two lines too
  ["Player"] = "Игрок",
  ["Cursor"] = "Курсор",
  
  thanks

**Comment by shagu** [comment]
Ooops! Updated: https://github.com/shagu/ShaguTweaks/commit/bad71dc665af202ad8cb4864ed10285f19f87da1

**Comment by shikulja** [comment]
How can I add a translation for ShaguTweaks-mods?
I connect the translation files in the same way:

ShaguTweaks_translation = ShaguTweaks_translation or {}
ShaguTweaks_translation["ruRU"] = {
....
}

but the translation is not detected since it looks up the table in ShaguTweaks (core addon)

**Comment by shagu** [comment]
Should be enough to do it like that:
```lua
ShaguTweaks_translation["ruRU"]["A Random Text"] = "случайный текст"
```

**Comment by shagu** [comment]
If you have more translations to do, you can also create a temporary table, that will then write its contents to the main table:

```lua
local ruRU = {
  ["A Random Text"] = "случайный текст",
  ["Another Text"] = "Другой текст",
}

for k, v in pairs(ruRU) do
  ShaguTweaks_translation["ruRU"][k] = v
end
```
